lib/providers/actionProviderComposer.js

```diff
@@ -0,0 +1,43 @@
+/**
+ * @name actionProviderComposer
+ * @description a function that satisfies the Atom IDE CodeActionProvider
+ * interface, running each diagnostic through a series of providers, each
+ * of which may return either a valid CodeAction (optionally wrapped in a
+ * Promise) as defined by Atom IDE or null if a given provider chooses not
+ * to provide an action for the diagnostic
+ * @param {JavaLanguageClient} languageClient the wrapping Java language client
+ * @param {[function]} providers a list of providers to register
+ * @returns {atom$CodeActionProvider}
+**/
+function actionProviderComposer(languageClient, ...providers) {
+  return {
+    grammarScopes: [ 'source.java' ],
+    priority: 1,
+    getCodeActions(editor, range, diagnostics) {
+      try {
+        return Promise.all(
+          diagnostics.reduce((acc, diagnostic) =>
+            acc.concat(
+              providers
+              .map(p => p(editor, range, diagnostic, languageClient))
+              .filter(action => !!action)
+              .reduce((_acc, action) => {
```

Comment:
Was looking over this extra reduction here. This extra flattening step was done because I think at some point the interface for a CodeAction allowed the `apply` fn to return an array of promises. Looking [back](https://github.com/facebook-atom/atom-ide-ui/blob/master/modules/atom-ide-ui/pkg/atom-ide-code-actions/lib/types.js) at it this no longer seems to be case. If we don't care about providers that return multiple actions for a single diagnostic then this can probably be removed.

Reply:
I think multiple actions for a diagnostic is quite a common case - certainly in C# so let's keep that as-is. 

I can merge this if you're ready - would love to be able to use up a more comprehensive java test project for my testing if you know of one that's available (my Java experience is a little dated)